[windows] properly compute programdata rather than hard-code to c:

derekwbrown opened the pull request:
### What does this PR do?

Properly computes the location of the `programdata` directory, instead of
assuming it's on `c:`

### Motivation

customer feedback

### Additional Notes

For completeness, will require forthcoming PRs for process & trace

derekwbrown committed:
Add timing statements to dep

derekwbrown committed:
[windows] properly compute programdata rather than hard-code to c:

trishankatdatadog reviewed (changes requested):
Thanks for the PR, Derek!

It LGTM for the most part. One issue I see is that if getting the Program Data directory fails for whatever, we fallback to the default value, and the user might run into problems anyway. But I don't see a good way around this right now.

derekwbrown committed:
convert string conversion to Cgo (at least for now) because the original
conversion doesn't pass `go vet`

derekwbrown commented on cmd/agent/app/integrations_windows.go:
```diff
@@ -24,6 +26,12 @@ var (
 	tufPipCachePath      = filepath.Join("c:\\", "ProgramData", "Datadog", "repositories", "cache")
 )
 
+func init() {
```
init() gets called on initialization, prior to `main`

derekwbrown commented on pkg/util/winutil/shutil.go:
```diff
@@ -0,0 +1,69 @@
+	err = SHGetKnownFolderPath(&FOLDERID_ProgramData, 0, 0, &retstr)
+	if err == nil {
+		// convert the string
+		defer CoTaskMemFree(retstr)
```
well, yess it's possible that it's non-nil, but if the error is set then retstr can't be counted on to be valid to do anything with, including free.

truthbk commented on cmd/agent/app/integrations_windows.go:
```diff
@@ -11,6 +11,8 @@ package app
+	"github.com/DataDog/datadog-agent/pkg/util/winutil"
 )
 
 const (
```
not really, no.

truthbk commented on pkg/util/winutil/shutil.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+// SHGetKnownFolderPath syscall to windows native SHGetKnownFOlderPath
+func SHGetKnownFolderPath(rfid *GUID, dwFlags uint32, hToken syscall.Handle, pszPath *uintptr) (retval error) {
```
We probably don't need to export this or `CoTaskMemFree` below beyond the `winutil` module, so I'd probably keep these as `sHGetKnownFolderPath` and `coTaskMemFree`.

derekwbrown replied:
It's being used in parallel PRs for `trace` & `process`. I think it has to be exported.

truthbk commented on pkg/util/winutil/shutil.go:
```diff
@@ -0,0 +1,73 @@
+	if err == nil {
+		// convert the string
+		defer CoTaskMemFree(retstrptr)
+		// go vet: "possible misuse of unsafe.Pointer"
```
can you provide some context?

derekwbrown replied:
Added to the comment.
IMO the original code would be better (b/c it doesn't require `cgo`), but I couldn't get it past CI.

truthbk replied:
I think something needs reformatting in `pkg/util/winutil/shutil.go`, CI is failing.

truthbk reviewed (commented):
This looks great for the most part, I'm not 100% sure about where we should initialize all of this stuff: `init()` blocks vs explicit initalization some other way.... The syscall look good, pointer safety looks good, memory management looks good too.

derekwbrown commented on cmd/agent/common/common_windows.go:
```diff
@@ -45,6 +46,15 @@ const (
 	DefaultDCALogFile = "c:\\programdata\\datadog\\logs\\cluster-agent.log"
 )
 
+func init() {
```
This was the minimum code change to make it work.  We could do some explicity initialization, but
1) this "should" never fail
2) in the event that it does, falling back to the default (which is right 99+% of the time) IMO is OK.
3) if it does fail, I would expect just about everything else to fail.

The other alternative would be to log to the event viewer, which we should still have access to here, if you're really concerned.

truthbk replied:
I think I would like to log something in the event of (2) - it's the kind of thing that might make us lose a ton of time debugging if we log nothing, whereas it'd provide a very useful datapoint if something strange happens here. I would prefer to log to our agent logs to be totally honest, but if that's problematic the event viewer might be better than nothing. Event viewer agent entries are not shipped with the flare, right?

derekwbrown committed:
review feedback

derekwbrown committed:
Move the definition of the service name to a central location
(pkg/config)

Create shared function for writing single event to event viewer.

In all cases that we can't find the ProgramData directory, log
to event viewer the failure

derekwbrown committed:
Merge branch 'db/deptiming' into db/programdata

derekwbrown committed:
Create shared function for writing single event to the event viewer

In all cases we can't find the programdata director, log to event viewer
the failure

truthbk commented on cmd/agent/app/integrations_windows.go:
```diff
@@ -30,6 +31,8 @@ func init() {
 	if err == nil {
 		tufPipCachePath = filepath.Join(pd, "Datadog", "repositories", "cache")
+	} else {
+		winutil.LogEventViewer(config.ServiceName, 0x8000000F, tufPipCachePath)
```
Not a huge fan of these magic numbers, but I can see we use them elsewhere. At some point we should maybe use consts and bit masks for the log level and message number. Can we create a card for that?

truthbk commented on pkg/util/winutil/eventlog.go:
```diff
@@ -0,0 +1,30 @@
+	switch msgnum & 0xF0000000 {
+	case 0x40000000:
+		// Info level message
+		elog.Info(msgnum, arg)
```
Just want to make sure...

is the `msgnum` we want to pass here the full message number, or do we have to mask out the most significant bits?

derekwbrown replied:
It's the full message number.  The high order 4 bits just indicates if it's info/warning/error level.  So the `case` is just so we send it to the right `log` method.  But, we need the whole message number.

truthbk reviewed (commented):
I think I'm satisfied with this now as far as I'm concerned. I'd like us to cleanup the magic number stuff, but in a different PR - no need to do it here.

Can we remove the `go dep` stuff and do that in a separate PR?

derekwbrown committed:
Back out unrelated change for showing `dep` timing

truthbk reviewed (dismissed):
Looks good to me! Thanks!
📦

derekwbrown committed:
Fix build error in systray

derekwbrown committed:
Fix ci issues
